File: solutions/2026-01-10/java/2026-01-10-minimum-ascii-delete-sum-for-two-strings.java
```java
// Problem: Minimum ASCII Delete Sum for Two Strings
// Link: https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/
//
// Approach:
// This problem can be solved using dynamic programming. We want to find the longest common subsequence (LCS)
// of the two strings, but instead of counting the length, we want to maximize the sum of ASCII values of
// the characters in the common subsequence. The reason for this is that the total ASCII sum of both strings
// minus twice the ASCII sum of the LCS will give us the minimum ASCII delete sum.
//
// Let dp[i][j] represent the maximum ASCII sum of a common subsequence of s1[0...i-1] and s2[0...j-1].
//
// Base cases:
// dp[0][j] = 0 for all j (empty s1, no common subsequence)
// dp[i][0] = 0 for all i (empty s2, no common subsequence)
//
// Recurrence relation:
// If s1[i-1] == s2[j-1]:
//   dp[i][j] = dp[i-1][j-1] + ascii(s1[i-1])
//   (We extend the common subsequence by including the matching character)
// If s1[i-1] != s2[j-1]:
//   dp[i][j] = max(dp[i-1][j], dp[i][j-1])
//   (We have to delete either s1[i-1] or s2[j-1], so we take the maximum sum from the previous states)
//
// After computing the dp table, the maximum ASCII sum of the LCS will be dp[s1.length()][s2.length()].
//
// The total ASCII sum of s1 is sum(s1).
// The total ASCII sum of s2 is sum(s2).
// The minimum delete sum = sum(s1) + sum(s2) - 2 * LCS_ASCII_SUM.
//
// Time complexity: O(m * n), where m is the length of s1 and n is the length of s2.
// We iterate through the dp table of size (m+1) x (n+1).
//
// Space complexity: O(m * n) for the dp table. This can be optimized to O(min(m, n)) by using only two rows
// of the dp table since each cell only depends on the previous row and the current row.
// However, for simplicity and clarity, we'll use the O(m*n) space approach first.
// The problem constraints (lengths up to 1000) make O(m*n) space acceptable.

class Solution {
    public int minimumDeleteSum(String s1, String s2) {
        int m = s1.length();
        int n = s2.length();

        // dp[i][j] will store the maximum ASCII sum of a common subsequence
        // for the first i characters of s1 and the first j characters of s2.
        int[][] dp = new int[m + 1][n + 1];

        // Fill the dp table
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                // If the current characters match
                if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                    // We extend the common subsequence by adding the ASCII value of the matching character.
                    dp[i][j] = dp[i - 1][j - 1] + s1.charAt(i - 1);
                } else {
                    // If the characters don't match, we consider two options:
                    // 1. Skip the current character from s1 (dp[i-1][j])
                    // 2. Skip the current character from s2 (dp[i][j-1])
                    // We take the maximum ASCII sum from these options.
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        // The maximum ASCII sum of the longest common subsequence is dp[m][n].
        int lcsAsciiSum = dp[m][n];

        // Calculate the total ASCII sum of both strings.
        int totalAsciiSum = 0;
        for (int i = 0; i < m; i++) {
            totalAsciiSum += s1.charAt(i);
        }
        for (int i = 0; i < n; i++) {
            totalAsciiSum += s2.charAt(i);
        }

        // The minimum delete sum is the total ASCII sum of both strings minus twice the ASCII sum of the LCS.
        // This is because characters in the LCS are not deleted from either string.
        // Total sum = (sum of deleted from s1) + (sum of deleted from s2) + 2 * (sum of LCS)
        // Minimum delete sum = (sum of deleted from s1) + (sum of deleted from s2)
        // So, Minimum delete sum = Total sum - 2 * (sum of LCS)
        return totalAsciiSum - 2 * lcsAsciiSum;
    }
}
```

File: solutions/2026-01-10/javascript/2026-01-10-minimum-ascii-delete-sum-for-two-strings.js
// Problem: Minimum ASCII Delete Sum for Two Strings
// Link: https://leetcode.com/problems/minimum-ascii-delete-sum-for-two-strings/
//
// Approach:
// This problem can be solved using dynamic programming. We want to find the longest common subsequence (LCS)
// in terms of ASCII values. The characters that are NOT part of this LCS are the ones we need to delete.
//
// Let dp[i][j] represent the maximum ASCII sum of a common subsequence of s1[0...i-1] and s2[0...j-1].
//
// Base Cases:
// dp[0][0] = 0 (empty strings have a common subsequence with sum 0)
// dp[i][0] = 0 for all i (s2 is empty, no common subsequence possible)
// dp[0][j] = 0 for all j (s1 is empty, no common subsequence possible)
//
// Recurrence Relation:
// If s1[i-1] == s2[j-1]:
//   The current characters match. We can include them in the common subsequence.
//   dp[i][j] = dp[i-1][j-1] + s1.charCodeAt(i-1)
// If s1[i-1] != s2[j-1]:
//   The current characters do not match. We have two choices:
//   1. Don't include s1[i-1] in the common subsequence (consider s1[0...i-2] and s2[0...j-1]).
//      The common subsequence sum would be dp[i-1][j].
//   2. Don't include s2[j-1] in the common subsequence (consider s1[0...i-1] and s2[0...j-2]).
//      The common subsequence sum would be dp[i][j-1].
//   We take the maximum of these two options to maximize the common subsequence sum.
//   dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1])
//
// After filling the DP table, dp[s1.length][s2.length] will hold the maximum ASCII sum of the common subsequence.
// The total ASCII sum of both strings is sum(s1) + sum(s2).
// The minimum ASCII delete sum will be (total ASCII sum) - 2 * (maximum common subsequence ASCII sum).
// We multiply by 2 because the characters in the LCS are present in both original strings, and they
// don't need to be deleted from either.
//
// Time Complexity: O(m * n), where m is the length of s1 and n is the length of s2.
//   We iterate through the entire DP table of size (m+1) * (n+1).
// Space Complexity: O(m * n) for the DP table.
//   This can be optimized to O(min(m, n)) by observing that dp[i][j] only depends on the previous row
//   or current row and previous column values. However, for clarity and direct implementation of the
//   recurrence, O(m*n) space is used here.

/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
var minimumDeleteSum = function(s1, s2) {
    const m = s1.length;
    const n = s2.length;

    // dp[i][j] will store the maximum ASCII sum of a common subsequence
    // of s1[0...i-1] and s2[0...j-1].
    const dp = Array(m + 1).fill(0).map(() => Array(n + 1).fill(0));

    // Calculate the maximum ASCII sum of the common subsequence
    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            // If the current characters match
            if (s1[i - 1] === s2[j - 1]) {
                // Include the current character's ASCII value in the common subsequence sum
                dp[i][j] = dp[i - 1][j - 1] + s1.charCodeAt(i - 1);
            } else {
                // If characters don't match, take the maximum sum from either
                // excluding s1[i-1] or excluding s2[j-1] from the common subsequence.
                dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // The total ASCII sum of s1 and s2
    let totalAsciiSum = 0;
    for (let i = 0; i < m; i++) {
        totalAsciiSum += s1.charCodeAt(i);
    }
    for (let i = 0; i < n; i++) {
        totalAsciiSum += s2.charCodeAt(i);
    }

    // The maximum ASCII sum of the common subsequence is stored in dp[m][n].
    // Characters in the common subsequence do not need to be deleted from either string.
    // So, we subtract twice the common subsequence sum from the total sum.
    const maxCommonSubsequenceAsciiSum = dp[m][n];
    const minDeleteSum = totalAsciiSum - 2 * maxCommonSubsequenceAsciiSum;

    return minDeleteSum;
};
```